sharktank/tests/models/llama4/llama4_test_iree_eager.py

```diff
@@ -0,0 +1,411 @@
+    L = inverted_mask.shape[-1]
+    causal = torch.tril(
+        torch.ones(L, L, dtype=inverted_mask.dtype, device=inverted_mask.device)
+    )
+    return causal.unsqueeze(0)  # (1, L, L)
+
+
+def deterministic_mask(inverted_mask):
+    seq_len = inverted_mask.shape[-1]
+    causal = torch.tril(
+        torch.ones(
+            seq_len, seq_len, dtype=inverted_mask.dtype, device=inverted_mask.device
+        )
+    )
+    return causal.unsqueeze(0)
+
+
+def deterministic_input_mask(seq_lens, sl, *, device="cpu"):
+    return torch.zeros(seq_lens.size(0), sl, dtype=torch.bool, device=device)
+
+
+def deterministic_page_table(page_count, config, dtype=torch.float16):
+
+    sub_page_dims = (
+        config.hp.attention_head_count_kv,  # 4
+        config.hp.expert_used_count,  # 2
+        config.hp.block_count,  # 4
+        config.block_seq_stride,  # 13
+        config.hp.attn_head_dim,  # 8
+    )
+    flat_dim = math.prod(sub_page_dims)
+    table = torch.arange(page_count * flat_dim, dtype=dtype).view(page_count, flat_dim)
+
+    return [table]
+
+
+def llama4_toy_sample_inputs(
+    model: PagedLlmModelV1,
+    config: LlamaModelConfig,
+    batch_size: int = 1,
+    dtype=torch.int64,
+) -> tuple[tuple[AnyTensor], OrderedDict[str, AnyTensor]]:
+
+    hp = config.hp
+    context_len = hp.context_length
+    base = torch.arange(context_len) % hp.vocab_size
+    tokens = base.repeat(batch_size, 1).long()
+
+    page_count = (len(tokens[0]) // config.block_seq_stride) * batch_size
+    kv_cache_state = deterministic_page_table(
+        page_count, config
+    )  # model.cache.allocate(page_count)
+
+    seq_block_ids = torch.arange(
+        start=0, end=tokens.numel() // config.block_seq_stride, dtype=dtype
+    ).view(batch_size, context_len // config.block_seq_stride)
+
+    hf_2d_attention_mask = torch.ones_like(tokens, dtype=dtype)
+    inverted_mask = hf_2d_attention_mask == 0
+    attention_mask = deterministic_mask(
+        inverted_mask
+    )  # model.attention_mask(inverted_mask)
+    seq_lens = torch.full((batch_size,), context_len, dtype=dtype)
+
+    args = tuple()
+    kwargs = OrderedDict(
+        (
+            ("tokens", tokens),
+            ("seq_lens", seq_lens),
+            ("attention_mask", [attention_mask]),
+            ("seq_block_ids", [seq_block_ids]),
+            ("cache_state", kv_cache_state),
+        )
+    )
+    return args, kwargs
+
+
+def patch_forward_moe(
+    self,
+    h: torch.Tensor | ShardedTensor,
+):
+    batch_size, sequence_length, feature_dim = h.shape
+    ffn_input = h.view(-1, feature_dim)
+
+    # For each token, the router calculates the router weights for all experts
+    # shape: (batch_size * sequence_length, expert_count)
+    router_logits = self.ffn_gate_inp(ffn_input)
+    router_weights = self.score_experts(router_logits.to(torch.float))
+
+    router_weights = reshard_like(router_weights, like=ffn_input)
+
+    def pick_first_expert(scores: torch.Tensor, k: int = 1):
```

Comment:
We shouldn't be using patched layers since we will no longer be testing the actual layers. Any change to the MoE block that breaks it won't be caught by this test.

Reply:
Yes, if there is something wrong or not implemented in the MoE, we should do that instead.